File: pset6/readability/readability.py
from cs50 import get_string

s = get_string("Text: ")

letters = 0
words = 0
sentences = 0

word_len = 0
for c in s:
    # ".", "!" or "?" indicates the end of a word and a sentence
    if word_len != 0 and c in [".", "!", "?"]:
        sentences += 1
        word_len = 0
    
    # The beginning of a word
    elif word_len == 0 and c.isalpha() == True:
        letters += 1
        words += 1
        word_len += 1
    
    # The rest of a word
    elif word_len != 0 and c.isalpha() == True:
        letters += 1
        word_len += 1
    
    # The end of a word
    elif word_len != 0 and c == " ":
        word_len = 0

grade = round(0.0588 * (letters / words) * 100 - 0.296 * (sentences / words) * 100 - 15.8)

if grade >= 16:
    print("Grade 16+")
elif grade < 1:
    print("Before Grade 1")
else:
    print(f"Grade {grade}")

File: pset2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int main(void)
{
    // Promp user for text
    string s = get_string("Text: ");
    int letters = 0, words = 1, sentences = 0;
    
    // Analyze text
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        if (isalpha(s[i]))
        {
            letters++;
        }
        else if (s[i] == ' ')
        {
            words++;
        }
        else if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            sentences++;
        }
    }
    
    // Prevent malicious users
    if (letters == 0)
    {
        words = 0;
        sentences = 0;
    }
    
    // Caculate index
    float L = (float) letters / (float) words * 100;
    float S = (float) sentences / (float) words * 100;
    int index = round(0.0588 * L - 0.296 * S - 15.8);
    
    // Print the result
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}